elasticdl/python/common/model_handler.py

```diff
@@ -61,57 +66,93 @@ def get_model_to_train(self, model):
-        Sequential and functional API model.
+        Restore Keras model definition by replacing elasticdl.layers.Embedding
+        layers with tf.keras.Embedding layers.
+        """
+        # clear keras model session to avoid clutter from old models/layers.
+        tf.keras.backend.clear_session()
+        if (
+            type(model) == tf.keras.models.Model
+            and not model._is_graph_network
+        ):
+            model = self._replace_embedding_attributes_for_subclass(
+                model, Embedding, tf.keras.layers.Embedding
+            )
+        else:
+            model = self._replace_embedding_layer_to_clone_model(
+                model, Embedding, tf.keras.layers.Embedding
+            )
+        return model
+
+    def _replace_embedding_layer_to_clone_model(
+        self, model, src_embedding_class, dst_embedding_class
+    ):
+        """
+        Clone a new model by cloning model and replace the
+        src_embedding_class layer with a dst_embedding_class.
         """
 
         def _clone_function(layer):
-            if type(layer) == tf.keras.layers.Embedding:
-                logger.info("Replace Keras Embedding with ElasticDL Embedding")
-                edl_embedding_layer = Embedding(
+            if type(layer) == src_embedding_class:
+                logger.info(
+                    "Replace {} with {}".format(
+                        src_embedding_class, dst_embedding_class
+                    )
+                )
+                embedding_layer = dst_embedding_class(
                     output_dim=layer.output_dim,
                     input_dim=layer.input_dim,
                     embeddings_initializer=layer.embeddings_initializer,
                     mask_zero=layer.mask_zero,
                     input_length=layer.input_length,
                 )
-                return edl_embedding_layer
+                return embedding_layer
             return layer
 
         return tf.keras.models.clone_model(
             model, clone_function=_clone_function
         )
 
-    def _replace_embedding_attributes_for_subclass(self, model):
+    def _replace_embedding_attributes_for_subclass(
+        self, model, src_embedding_class, dst_embedding_class
+    ):
         """
         Replace the keras embedding attribute with
         elasticdl.layers.Embedding layer.
         """
-        for attr_name, attr_value in model.__dict__.items():
-            if type(attr_value) == tf.keras.layers.Embedding:
-                edl_embedding_layer = Embedding(
-                    output_dim=attr_value.output_dim,
-                    input_dim=attr_value.input_dim,
-                    embedding_initializer=attr_value.embeddings_initializer,
-                    mask_zero=attr_value.mask_zero,
-                    input_length=attr_value.input_length,
+        for name, value in model.__dict__.items():
+            if type(value) == src_embedding_class:
```

Comment:
use `isinstance` instead of string comparison?

Reply:
`isinstance(value, class)` returns True if the value is a subclass instance of the class. Here, we only want to replace `tf.keras.layers.Embedding` and don't replace any custom layer extended `tf.keras.layers.Embedding`. So, I use `type` and not `isinstance`.